cmd/zfs/zfs_main.c

```diff
@@ -984,24 +1003,31 @@ zfs_do_create(int argc, char **argv)
 			*p = '/';
 		if (zpool_handle == NULL)
 			goto error;
-		spa_version = zpool_get_prop_int(zpool_handle,
-		    ZPOOL_PROP_VERSION, NULL);
-		if (spa_version >= SPA_VERSION_REFRESERVATION)
-			resv_prop = ZFS_PROP_REFRESERVATION;
-		else
-			resv_prop = ZFS_PROP_RESERVATION;
 
 		(void) snprintf(msg, sizeof (msg),
+		    dryrun ? gettext("cannot verify '%s'") :
 		    gettext("cannot create '%s'"), argv[0]);
 		if (props && (real_props = zfs_valid_proplist(g_zfs, type,
 		    props, 0, NULL, zpool_handle, B_TRUE, msg)) == NULL) {
 			zpool_close(zpool_handle);
 			goto error;
 		}
-		zpool_close(zpool_handle);
+	}
 
-		volsize = zvol_volsize_to_reservation(volsize, real_props);
-		nvlist_free(real_props);
+	if (type == ZFS_TYPE_VOLUME && !noreserve) {
+		uint64_t spa_version;
+		zfs_prop_t resv_prop;
+		char *strval;
+
+		spa_version = zpool_get_prop_int(zpool_handle,
+		    ZPOOL_PROP_VERSION, NULL);
+		if (spa_version >= SPA_VERSION_REFRESERVATION)
+			resv_prop = ZFS_PROP_REFRESERVATION;
+		else
+			resv_prop = ZFS_PROP_RESERVATION;
+
+		volsize = zvol_volsize_to_reservation(zpool_handle, volsize,
+		    real_props);
```

Comment:
While you're here, there's a long standing related issue https://github.com/zfsonlinux/zfs/issues/8541#issue-426402842 that it would be good to address.  The actual behavior does not match the man page description.

Reply:
There are other issues in nearby code too, but I'm not sure that issues are filed for all of them.

- If `copies` is not specified it is assumed to be 1.  That assumption is wrong when nearest ancestor dataset has `copies` != 1.
- `zfs create -V $size -o refreservation=auto` should not try to a volume with size `(uint64_t)-1`.  I haven't verified that exists in zol, but certainly an issue in illumos.

I'd prefer to deal with those each on their own.